llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+    // do not have to be updated
+    AU.addRequired<LiveDebugVariables>();
+    AU.addPreserved<LiveDebugVariables>();
+    AU.addRequired<LiveStacks>();
+    AU.addPreserved<LiveStacks>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  MachineFunction *MF = nullptr;
+  const MachineRegisterInfo *MRI = nullptr;
+  const AIEBaseRegisterInfo *TRI = nullptr;
+  VirtRegMap *VRM = nullptr;
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
+
+  /// Get all the defined physical registers that the MachineBasicBlock already
+  /// uses. These physical registers should not be used for replacement
+  /// candidates, since this would introduce new WAW dependencies, which this
+  /// pass tries to remove.
+  BitVector getDefinedPhysRegs(const MachineBasicBlock *MBB) const;
+
+  bool replaceReg(const Register Reg, BitVector &BlockedPhysRegs);
+
+  /// Find a free register of the same register class type, but
+  /// exclude the blocked physical registers from the result.
+  /// Otherwise a new WAW dependencies can be introduced, that was previously
+  /// removed.
+  MCPhysReg getReplacementPhysReg(const Register Reg,
+                                  const BitVector &BlockedPhysRegs) const;
+
+  bool isWorthRenaming(const Register &Reg, const BitVector &UsedPhysRegs,
+                       const SmallVector<Register, 16> &VRegWithCopies) const;
+
+  /// return the Physical register of the machine operand.
+  MCPhysReg getAssignedPhysReg(const MachineOperand &MO) const;
+
+  bool isIdentityCopy(const MachineInstr &MI) const;
+
+  /// MachineInstructions can write into sub-registers (%0:sub_256_lo and
+  /// %0:sub_256_hi). In the VirtualRegisterMap, they block the same
+  /// register (i.e. $x0). This can cause issues, when detecting Write
+  /// After Write (WAW) dependencies. To mitigate this, only use the last
+  /// definition of a virtual register to count the definitions, so that
+  /// writing into subregisters does not already trigger the register
+  /// replacement. In this case $x0 would have already been replaced, even
+  /// though there is no real WAW dependency.
+  ///                                                                          \
+  /// undef %0.sub_256_lo:mxa, _, _= VLDA_2D_dmw_lda_w _,_                     \
+  /// % 0.sub_256_hi:vec512 = VLDA_dmw_lda_w_ag_idx_imm _,_                    \
+  /// %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  ///                                                                          \
+  /// returns a mapping between a virtual register and its last defining machine
+  /// instruction.
+  std::map<Register, const MachineInstr *>
```

Comment:
FYI, you could use a densely packed map (`IndexedMap` for instance) by leveraging the Register::virtReg2Index helper method.

Reply:
updated the data structure